presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/StaticMetastoreLocator.java

```diff
@@ -76,23 +82,43 @@ public StaticMetastoreLocator(List<URI> metastoreUris, @Nullable String metastor
     public ThriftMetastoreClient createMetastoreClient(Optional<String> delegationToken)
             throws TException
     {
-        List<HostAndPort> metastores = new ArrayList<>(addresses);
-        Collections.shuffle(metastores.subList(1, metastores.size()));
+        List<Address> sortedAddresses = addresses.stream()
+                .sorted(Comparator.comparingLong(address -> address.getBackoff().getBackoffDuration()))
+                .collect(toImmutableList());
 
         TException lastException = null;
-        for (HostAndPort metastore : metastores) {
+        for (Address address : sortedAddresses) {
             try {
-                ThriftMetastoreClient client = clientFactory.create(metastore, delegationToken);
-                if (!isNullOrEmpty(metastoreUsername)) {
-                    client.setUGI(metastoreUsername);
-                }
-                return client;
+                return getClient(address, delegationToken);
             }
             catch (TException e) {
                 lastException = e;
             }
         }
-        throw new TException("Failed connecting to Hive metastore: " + addresses, lastException);
+        throw new TException("Failed connecting to Hive metastore: " + addresses.stream().map(Address::getMetastore).collect(toImmutableList()), lastException);
+    }
+
+    private ThriftMetastoreClient getClient(Address address, Optional<String> delegationToken)
+            throws TException
+    {
+        ThriftMetastoreClient client = new FailureAwareThriftMetastoreClient(clientFactory.create(address.getMetastore(), delegationToken), new Callback()
```

Comment:
Imagine that user creates a table and HMS fails? Can we retry that operation? It is not known why HMS failed maybe it managed to create a table but failed because of something else? In general this applies to every method that modifies the state of HMS.

Maybe we should only retry only on reading path?

It looks like acquiring the lock also modifies the state of HMS so we should not retry that. Hence this should be disabled at all for Hive ACID.

Reply:
What retry do you mean here? retry logic happens in `ThriftHiveMetastore` right now. Do you want me to change the code there? Right now retry will happen on same HMS. for example, `createTable`
```
  @Override
    public void createTable(HiveIdentity identity, Table table)
    {
        try {
            retry()
                    .stopOn(AlreadyExistsException.class, InvalidObjectException.class, MetaException.class, NoSuchObjectException.class)
                    .stopOnIllegalExceptions()
                    .run("createTable", stats.getCreateTable().wrap(() -> {
                        try (ThriftMetastoreClient client = createMetastoreClient(identity)) {
                            client.createTable(table);
                        }
                        return null;
                    }));
        }
        catch (AlreadyExistsException e) {
            throw new TableAlreadyExistsException(new SchemaTableName(table.getDbName(), table.getTableName()));
        }
        catch (NoSuchObjectException e) {
            throw new SchemaNotFoundException(table.getDbName());
        }
        catch (TException e) {
            throw new PrestoException(HIVE_METASTORE_ERROR, e);
        }
        catch (Exception e) {
            throw propagate(e);
        }
    }
```
it will retry on `TException`.

Reply:
Right, thank you for explanation.

Reply:
I guess the conclusion is no action needed?